### PR TITLE
1303 war games action 1 message wizard enhancements

### PR DESCRIPTION
**Pull request by LukeLoze1**
# Checklist
Reminder: If any change made to this repo affects the acceptance tests then the version pin in the Pipfile in [census-rm-acceptance-tests]() needs to be updated.
* [ ] Updated census-rm-acceptance-tests version pin (if applicable)

# Motivation and Context
Enhance msgwizard

# What has changed
Pagination, filtering, mass quarantine and some other stuff 

# How to test?
Manually

### CODE REVIEW

**Comment by NickGrantX** [comment]
I did crash it, by accident:

```
Please enter the page you would like to see 1 - 1 or ENTER to exit: ^[[A^[[B
 is not a valid integer
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/toolbox/toolbox/message_tools/bad_message_wizard.py", line 598, in <module>
    main()
  File "/home/toolbox/toolbox/message_tools/bad_message_wizard.py", line 22, in main
    show_home_page()
  File "/home/toolbox/toolbox/message_tools/bad_message_wizard.py", line 51, in show_home_page
    actions[valid_selection - 1]['action']()
  File "/home/toolbox/toolbox/message_tools/bad_message_wizard.py", line 214, in filter_bad_messages
    paginate_messages(bad_message_summaries)
  File "/home/toolbox/toolbox/message_tools/bad_message_wizard.py", line 379, in paginate_messages
    start_index = (int(page_num) * ITEMS_PER_PAGE) - ITEMS_PER_PAGE
ValueError: invalid literal for int() with base 10: '\x1b[A\x1b[B'
```

**Review by NickGrantX — CHANGES_REQUESTED** [review]
Bug with quarantine all:

```
There are currently 1 bad messages:

        | Hash | Exception Message                        | First Seen          | Queues
     ---|-------------------------------------------------|---------------------|-------------------------
      1.| aadd | Questionnaire Id '1110000009' not found! | 2020-12-09T08:06:04 | case.questionnairelinked

Select a message (1 to 1), Quarantine all with q or press ENTER for page selection:


Actions:
  1. List bad messages
  2. Get bad message from hash
  3. Get quarantined messages
  4. Quarantine all bad messages
  5. Reset bad message cache
  6. Filter bad messages

Choose an action: 4
There are currently 30 bad messages, continuing will quarantine them all

1. Continue
2. Cancel

Choose an action:
```

**Comment by NickGrantX** [comment]
It gets worse (kinda)....

```
Actions:
  1. List bad messages
  2. Get bad message from hash
  3. Get quarantined messages
  4. Quarantine all bad messages
  5. Reset bad message cache
  6. Filter bad messages

Choose an action: 1

There are currently no bad messages known to the exception manager


Actions:
  1. List bad messages
  2. Get bad message from hash
  3. Get quarantined messages
  4. Quarantine all bad messages
  5. Reset bad message cache
  6. Filter bad messages

Choose an action: 4
There are currently 30 bad messages, continuing will quarantine them all

1. Continue
2. Cancel

Choose an action: 1

Confirm by responding with 'I confirm I wish to quarantine all 30 bad messages' exactly: I confirm I wish to quarantine all 30 bad messages

Confirmed, quarantining all 30 messages
About to quarantine this number of msgs: 30
About to quarantine message:  3e5dbe1162279f72523e159416e812eb1156fa563638b1e82a283d8a2c543a14
About to quarantine message:  8aaa68bf8dc83617738103f6e735fb2e6b88da7d1402d23423706d88ff084daf
About to quarantine message:  f60843c36805faca75f44c2a32fdde735bbf7dce2d03155d594c56ba1d783c76
About to quarantine message:  4fa7be6ac03ed3bdf51d768ca13763bac5ebf74a84f8f5f4d938c29c015a9cb9
About to quarantine message:  96c3a050c288dbd7512bb1288ebc459d9c7247c806da1bfcea89d18201c80882
About to quarantine message:  d1a3e134bde0d10641f1074a14ef83eccff70a46b347c84bdd22cf14aae241d3
About to quarantine message:  c02c9e688f41f853af48152121534887743889902002be0c20c109b76dd8a59c
About to quarantine message:  da77d1960584182442c7e62d7d8f903ef5f66c2534bca0f353830cac9d4b07c6
About to quarantine message:  32d9a4fb032230df5f5bec4d0d5f6a929735f6cd9997e42c8dc9188b6b9e87a0
About to quarantine message:  aadd1b7ae60532f711c2ca42ac49b1cea303bec66a755096bfd58c1723650e0e
About to quarantine message:  32450d8a06a309b391e952f72d0cae130a654845ba1eba0889b5db6e98547940
About to quarantine message:  268f4783f22e4e9faf747312e8bcffa90fa3e08bc96eb5d177d443f0fa5c3e3d
About to quarantine message:  6328c7914bfac473596838e12303f70aa657981f8fafc703a8d0ab15537eef09
About to quarantine message:  ee15ad12ab2bb8b79cf76f03a1904baeb43d2811dfb4a554cf73a0382b4ad874
About to quarantine message:  49742b4b8dd3a7ff2a2a32410e34f55a57d09a2327edb26d726a30e400960966
About to quarantine message:  4ca1292eba082a40b64ff2e393194ace6c70a4fd135230eaf5fec3700f2b4073
About to quarantine message:  13b85b4c7297c7c4164647ef0ae1fa386db67a7acde9143dcf998405dd803e04
About to quarantine message:  9c6b06b00ac4c7a2cde78bcc10f65b39a87d33768cedf1e6536f603bd02941bf
About to quarantine message:  25717e3586570ff84b659370918ca5d63cc00e145f4d84a56e891d48b06a6c49
About to quarantine message:  6b0638007bd640fcf948f67240c10dcad9f918f0757162388b469552ce198e9d
About to quarantine message:  3be0ef95170d224139edfb9c22064cda907b4180069c0888c8beba8a59b6e8a2
About to quarantine message:  a73020bfe0253c456e773f2b981ad368c91a043dafa5feb33b33d0b0cdf057c4
About to quarantine message:  bd046745378f9aa867722154f5397666ebc197cecc558987111ab5970ec67164
About to quarantine message:  557ea88d9981ae247b851b80317350137e63d6ef936dd9fe37620d7397c45205
About to quarantine message:  acdd7ab8268651bf118b55e4412280624dc0ff953184f11eae976557c6875eed
About to quarantine message:  1cfc1c278755e1a4b45d9d753c43fd50645e7b131f66d37eef69da40558f2878
About to quarantine message:  389b5c1da3e7205c7d4e8f7bad08d37bb7d1277e2a55ad36cb552669c9226de6
About to quarantine message:  c5beba6dd9a4d5635f940618c885818b5abf0fc837349b9271f653abb0b05737
About to quarantine message:  f73f0c8986acf230c2f3d758497dae35ba5a4789496e261ddbcbd00fa5d9500b
About to quarantine message:  e2b24fb2e639196b28720a58d8e44890fbbc5b2be1916a97302f82ed031ffc08
Successfully quarantined all bad messages
```

**Comment by NickGrantX** [comment]
Even after I quarantined the "phantom" 30 bad messages, they stayed there: quarantine all kept showing there were 30 to quarantine  :(

File: toolbox/message_tools/bad_message_wizard.py [hunk 1]
```diff
@@ -156,37 +183,202 @@ def reset_bad_message_cache():
...
+
+
+def get_bad_message_summaries(all_message_summaries):
+    bad_message_summaries = [summary for summary in all_message_summaries if not summary['quarantined']]
+    return remove_milliseconds_from_firstseen(bad_message_summaries)
+
+
+def remove_milliseconds_from_firstseen(bad_message_summaries):
+    for msg in bad_message_summaries:
+        msg['firstSeen'] = msg['firstSeen'][0:len(msg['firstSeen']) - MILLISECOND_STR_LEN]
```
Comment (liamtoozer) [inline toolbox/message_tools/bad_message_wizard.py]: `msg['firstSeen'] = rfc3339.parse_datetime(msg['firstSeen']).strftime("%Y-%m-%dT%H:%M:%S")` will work if you wanted to shorten and get rid of an extra constant

File: toolbox/message_tools/bad_message_wizard.py [hunk 2]
```diff
@@ -156,37 +183,202 @@ def reset_bad_message_cache():
...
+    filtered_msgs = []
+
+    for msg in all_message_summaries:
+        message_hash = msg['messageHash']
+        message_metadata = get_bad_message_metadata(message_hash)
+
+        for (k, v) in message_metadata[0].items():
+            for item in v.values():
+                if re.search(filter_by_text.lower(), str(item).lower()):
+                    if msg not in filtered_msgs:
```
Comment (liamtoozer) [inline toolbox/message_tools/bad_message_wizard.py]: This `if` can be merged with above to save any extra nesting

File: toolbox/message_tools/bad_message_wizard.py [hunk 3]
```diff
@@ -156,37 +183,202 @@ def reset_bad_message_cache():
...
+
+    bad_message_summaries = get_bad_message_summaries(filtered_message_summaries)
+    if not bad_message_summaries:
+        show_no_bad_messages()
+        return
+
+    print(f'There are currently {len(bad_message_summaries)} bad messages matching your filter:')
+    paginate_messages(bad_message_summaries)
+
+    return
```
Comment (liamtoozer) [inline toolbox/message_tools/bad_message_wizard.py]: Empty return can be removed

**Review by liamtoozer — COMMENTED** [review]
A few cosmetic bits. I'm getting the same error as Nick, i.e. the quarantine doesn't appear to work

**Review by liamtoozer — APPROVED** [review]
Tested again locally with a few bad messages. LGTM 👍

**Comment by LukeLoze1** [comment]
> Bug with quarantine all:
> 
> ```
> There are currently 1 bad messages:
> 
>         | Hash | Exception Message                        | First Seen          | Queues
>      ---|-------------------------------------------------|---------------------|-------------------------
>       1.| aadd | Questionnaire Id '1110000009' not found! | 2020-12-09T08:06:04 | case.questionnairelinked
> 
> Select a message (1 to 1), Quarantine all with q or press ENTER for page selection:
> 
> 
> Actions:
>   1. List bad messages
>   2. Get bad message from hash
>   3. Get quarantined messages
>   4. Quarantine all bad messages
>   5. Reset bad message cache
>   6. Filter bad messages
> 
> Choose an action: 4
> There are currently 30 bad messages, continuing will quarantine them all
> 
> 1. Continue
> 2. Cancel
> 
> Choose an action:
> ```



> Even after I quarantined the "phantom" 30 bad messages, they stayed there: quarantine all kept showing there were 30 to quarantine :(

We just this with the old toolbox and got the same result and quarantine all wasn't filtering out previously quarantined msgs, this is hopefully fixed now

**Comment by LukeLoze1** [comment]
> It gets worse (kinda)....
> 
> ```
> Actions:
>   1. List bad messages
>   2. Get bad message from hash
>   3. Get quarantined messages
>   4. Quarantine all bad messages
>   5. Reset bad message cache
>   6. Filter bad messages
> 
> Choose an action: 1
> 
> There are currently no bad messages known to the exception manager
> 
> 
> Actions:
>   1. List bad messages
>   2. Get bad message from hash
>   3. Get quarantined messages
>   4. Quarantine all bad messages
>   5. Reset bad message cache
>   6. Filter bad messages
> 
> Choose an action: 4
> There are currently 30 bad messages, continuing will quarantine them all
> 
> 1. Continue
> 2. Cancel
> 
> Choose an action: 1
> 
> Confirm by responding with 'I confirm I wish to quarantine all 30 bad messages' exactly: I confirm I wish to quarantine all 30 bad messages
> 
> Confirmed, quarantining all 30 messages
> About to quarantine this number of msgs: 30
> About to quarantine message:  3e5dbe1162279f72523e159416e812eb1156fa563638b1e82a283d8a2c543a14
> About to quarantine message:  8aaa68bf8dc83617738103f6e735fb2e6b88da7d1402d23423706d88ff084daf
> About to quarantine message:  f60843c36805faca75f44c2a32fdde735bbf7dce2d03155d594c56ba1d783c76
> About to quarantine message:  4fa7be6ac03ed3bdf51d768ca13763bac5ebf74a84f8f5f4d938c29c015a9cb9
> About to quarantine message:  96c3a050c288dbd7512bb1288ebc459d9c7247c806da1bfcea89d18201c80882
> About to quarantine message:  d1a3e134bde0d10641f1074a14ef83eccff70a46b347c84bdd22cf14aae241d3
> About to quarantine message:  c02c9e688f41f853af48152121534887743889902002be0c20c109b76dd8a59c
> About to quarantine message:  da77d1960584182442c7e62d7d8f903ef5f66c2534bca0f353830cac9d4b07c6
> About to quarantine message:  32d9a4fb032230df5f5bec4d0d5f6a929735f6cd9997e42c8dc9188b6b9e87a0
> About to quarantine message:  aadd1b7ae60532f711c2ca42ac49b1cea303bec66a755096bfd58c1723650e0e
> About to quarantine message:  32450d8a06a309b391e952f72d0cae130a654845ba1eba0889b5db6e98547940
> About to quarantine message:  268f4783f22e4e9faf747312e8bcffa90fa3e08bc96eb5d177d443f0fa5c3e3d
> About to quarantine message:  6328c7914bfac473596838e12303f70aa657981f8fafc703a8d0ab15537eef09
> About to quarantine message:  ee15ad12ab2bb8b79cf76f03a1904baeb43d2811dfb4a554cf73a0382b4ad874
> About to quarantine message:  49742b4b8dd3a7ff2a2a32410e34f55a57d09a2327edb26d726a30e400960966
> About to quarantine message:  4ca1292eba082a40b64ff2e393194ace6c70a4fd135230eaf5fec3700f2b4073
> About to quarantine message:  13b85b4c7297c7c4164647ef0ae1fa386db67a7acde9143dcf998405dd803e04
> About to quarantine message:  9c6b06b00ac4c7a2cde78bcc10f65b39a87d33768cedf1e6536f603bd02941bf
> About to quarantine message:  25717e3586570ff84b659370918ca5d63cc00e145f4d84a56e891d48b06a6c49
> About to quarantine message:  6b0638007bd640fcf948f67240c10dcad9f918f0757162388b469552ce198e9d
> About to quarantine message:  3be0ef95170d224139edfb9c22064cda907b4180069c0888c8beba8a59b6e8a2
> About to quarantine message:  a73020bfe0253c456e773f2b981ad368c91a043dafa5feb33b33d0b0cdf057c4
> About to quarantine message:  bd046745378f9aa867722154f5397666ebc197cecc558987111ab5970ec67164
> About to quarantine message:  557ea88d9981ae247b851b80317350137e63d6ef936dd9fe37620d7397c45205
> About to quarantine message:  acdd7ab8268651bf118b55e4412280624dc0ff953184f11eae976557c6875eed
> About to quarantine message:  1cfc1c278755e1a4b45d9d753c43fd50645e7b131f66d37eef69da40558f2878
> About to quarantine message:  389b5c1da3e7205c7d4e8f7bad08d37bb7d1277e2a55ad36cb552669c9226de6
> About to quarantine message:  c5beba6dd9a4d5635f940618c885818b5abf0fc837349b9271f653abb0b05737
> About to quarantine message:  f73f0c8986acf230c2f3d758497dae35ba5a4789496e261ddbcbd00fa5d9500b
> About to quarantine message:  e2b24fb2e639196b28720a58d8e44890fbbc5b2be1916a97302f82ed031ffc08
> Successfully quarantined all bad messages
> ```

Previous issue with quarantine all, hopefully fixed now

**Comment by NickGrantX** [comment]
NullPointer exceptions are quite hard to track down, from first glance (the filter does work though)

```
Filter By Text: null
There are currently 6 bad messages matching your filter:
You are on page 1 of 1

        | Hash | Exception Message | First Seen          | Queues
     ---|-------------|---------------------|---------------
      1.| 140e | None | 2020-12-11T07:12:12 | Case.Responses
      2.| 5d06 | None | 2020-12-11T07:12:01 | Case.Responses
      3.| eef6 | None | 2020-12-11T07:12:19 | Case.Responses
      4.| a35a | None | 2020-12-11T07:12:36 | Case.Responses
      5.| 057b | None | 2020-12-11T07:15:41 | case.refusals
      6.| fbbd | None | 2020-12-11T07:15:58 | case.refusals

Select a message (1 to 6), Quarantine all with q or press ENTER for page selection: 1

-------------------------------------------------------------------------------------
Message Hash: 140e47e911e74b565b4628679b549eef54d6863902faadbb455ccb65ec121f02
Reports:
  1:
    Exception Report:
      messageHash: 140e47e911e74b565b4628679b549eef54d6863902faadbb455ccb65ec121f02
      service: Case Processor
      queue: Case.Responses
      exceptionClass: java.lang.NullPointerException
      exceptionMessage: None
      exceptionRootCause: 	at uk.gov.ons.census.casesvc.service.QidReceiptService.processReceipt(QidReceiptService.java:39)
    Stats:
      firstSeen: 2020-12-11T07:12:12.352996Z
      lastSeen: 2020-12-11T07:27:49.129098Z
      seenCount: 56
      loggedAtLeastOnce: True
-------------------------------------------------------------------------------------

Current message hash: 140e47e911e74b565b4628679b549eef54d6863902faadbb455ccb65ec121f02
```

**Review by NickGrantX — APPROVED** [review]
Works for me. Great work - LOADS of cool features there!

**Pull request merged**